File: Source/Shelfinator.Runner/SegmentItem.cpp
#include "SegmentItem.h"

#include <math.h>
#include "Helpers.h"

namespace Shelfinator
{
	namespace Runner
	{
		namespace SongData
		{
			void SegmentItem::Read(BufferFile::ptr file)
			{
				segmentIndex = file->GetInt();
				segmentStartTime = file->GetInt();
				segmentEndTime = file->GetInt();
				auto segmentTime = file->GetInt();
				startTime = file->GetInt();
				endTime = file->GetInt();
				auto startVelocity = file->GetInt();
				auto endVelocity = file->GetInt();
				auto baseVelocity = file->GetInt();

				auto time = endTime - startTime;

				v0 = (double)startVelocity / baseVelocity;
				auto v1 = (double)endVelocity / baseVelocity;
				a = 120 * segmentTime / pow(time, 5) - 60 * v0 / pow(time, 4) - 60 * v1 / pow(time, 4);
				b = -180 * segmentTime / pow(time, 4) + 96 * v0 / pow(time, 3) + 84 * v1 / pow(time, 3);
				c = 60 * segmentTime / pow(time, 3) - 36 * v0 / pow(time, 2) - 24 * v1 / pow(time, 2);
				mult = segmentEndTime >= segmentStartTime ? 1 : -1;
			}

			double SegmentItem::GetSegmentTime(int time)
			{
				if (segmentEndTime == segmentStartTime)
					return segmentEndTime;

				auto useTime = (double)time - startTime;
				auto value = std::fmod(a * pow(useTime, 5) / 20 + b * pow(useTime, 4) / 12 + c * pow(useTime, 3) / 6 + v0 * useTime, segmentEndTime - segmentStartTime) * mult + segmentStartTime;
				return value;
			}
		}
	}
}

File: Source/Shelfinator.Creator/SongData/SegmentItem.cs
using System.IO;

namespace Shelfinator.Creator.SongData
{
	class SegmentItem
	{
		public int SegmentIndex { get; set; }
		public int SegmentStartTime { get; set; }
		public int SegmentEndTime { get; set; }
		public int SegmentTime { get; set; }
		public int StartTime { get; set; }
		public int EndTime { get; set; }
		public int StartVelocity { get; set; }
		public int EndVelocity { get; set; }
		public int BaseVelocity { get; set; }
		public int Duration => EndTime - StartTime;

		public SegmentItem(int segmentIndex, int segmentStartTime, int segmentEndTime, int segmentTime, int startTime, int endTime, int startVelocity, int endVelocity, int baseVelocity)
		{
			SegmentIndex = segmentIndex;
			SegmentStartTime = segmentStartTime;
			SegmentEndTime = segmentEndTime;
			SegmentTime = segmentTime;
			StartTime = startTime;
			EndTime = endTime;
			StartVelocity = startVelocity;
			EndVelocity = endVelocity;
			BaseVelocity = baseVelocity;
		}

		public void Save(BinaryWriter output)
		{
			output.Write(SegmentIndex);
			output.Write(SegmentStartTime);
			output.Write(SegmentEndTime);
			output.Write(SegmentTime);
			output.Write(StartTime);
			output.Write(EndTime);
			output.Write(StartVelocity);
			output.Write(EndVelocity);
			output.Write(BaseVelocity);
		}
	}
}
